### PR TITLE
Allow searching by Issue Index

### PR DESCRIPTION
**Pull request by aceArt-GmbH**
Hey,
first of all thanks for creating and maintaining this repo :D

A user of our instance was trying to search for an issue via ID. Currently this does not work.
The alternative we have been using is opening any issue and changing the URL to the desired ID, but this workflow is not good for non technical users. (i.e. `/go-gitea/gitea/issues/17927` -> replace 17927 with whatever ticket ID I want to search)

I came up with this simple solution, but I am not sure if it is good enough.
For example, it does not allow searching for issues starting with the keyword (search "1" and find issue 12 and 13)
Please let me know about your opinion and an alternative implementation :)

image
issue `#3` is found although neither title nor content contains a `3`

### CODE REVIEW

File: modules/indexer/issues/indexer.go [hunk 1]
```diff
@@ -347,5 +348,9 @@ func SearchIssuesByKeyword(repoIDs []int64, keyword string) ([]int64, error) {
 	for _, r := range res.Hits {
 		issueIDs = append(issueIDs, r.ID)
 	}
+	i, err := strconv.ParseInt(keyword, 10, 64)
+	if err == nil {
+		issueIDs = append(issueIDs, i)
```
Comment (wxiaoguang) [inline modules/indexer/issues/indexer.go]: I think there are some problems:

1. `i` is not suitable for a formal variable name
2. Do you mean `issueID` or `issueIndex`?
3. If we do not use `repoIDs` to filter issues, the unrelated issues from other repositories would appear in result.

Comment (aceArt-GmbH) [reply modules/indexer/issues/indexer.go]: 1. agreed. Which name would be better? `keywordAsInt`?
2. what is `issueIndex`? I want to add an ID to the found IDs
3. My use-case is for the repo search, so the repository is known and it does not show issues from any other repo. Do you think the API could return something wrong?
`r.ID` from line 348 is just a issue number too and not the database ID.

Comment (wxiaoguang) [reply modules/indexer/issues/indexer.go]: `2` If you see http://domain/org/repo/issues/123, the 123 is the issue index. There can also be another issue with index 123 in another repo: http://domain/org/repo-other/issues/123 .  Issue ID is the unique internal auto-increment database column for each issue, users seldom see the issue IDs, instead, users use issue index everywhere. Of course, Gitea use issue ID internally everywhere.

`3` I believe the `r.ID` is the unique issue ID, not issue index. However, if there is another query filter outside `SearchIssuesByKeyword` to filter the issue IDs with repositories again, the final result can be correct, but it is still not safe to return a non-related issue from SearchIssuesByKeyword.

Comment (wxiaoguang) [reply modules/indexer/issues/indexer.go]: In my mind: an easy way to find issues by the indexes can be `put the issue indexes into the search indexes (make the issue indexes searchable)`, then we do not need to change SearchIssuesByKeyword, then the issue indexes can be found correctly.

Comment (aceArt-GmbH) [reply modules/indexer/issues/indexer.go]: oh I see, it worked only because of my fresh database where issueIndex was also issueID :/

**Comment by lunny** [comment]
I would like to input `#3` and enter so that we can just open the issue directly.

File: modules/indexer/issues/indexer.go (hunk 1 above)
Comment (aceArt-GmbH) [reply modules/indexer/issues/indexer.go]: >put the issue indexes into the search indexes (make the issue indexes searchable)

I am not familiar enough with the code base to know how to do this

**Comment by aceArt-GmbH** [comment]
we had that idea too, but is it intuitive?
For you probably as you know `#3` references issues, but less technical users maybe not

File: modules/indexer/issues/indexer.go (hunk 1 above)
Comment (wxiaoguang) [reply modules/indexer/issues/indexer.go]: You can take a look at `IndexerData`

```go
// IndexerData data stored in the issue indexer
type IndexerData struct {
	ID       int64    `json:"id"`
	RepoID   int64    `json:"repo_id"`
	Title    string   `json:"title"`
	Content  string   `json:"content"`
	Comments []string `json:"comments"`
	IsDelete bool     `json:"is_delete"`
	IDs      []int64  `json:"ids"`
}
```

Follow the way how `Content` are put into search indexer.

And then make the search engine use them: 

```go
indexerQuery := bleve.NewConjunctionQuery( .... newMatchPhraseQuery(keyword, "Comments", issueIndexerAnalyzer), ...
kwQuery := elastic.NewMultiMatchQuery(keyword, "title", "content", "comments", ....)
```

Comment (aceArt-GmbH) [reply modules/indexer/issues/indexer.go]: I feel like I am close, yet it won't work. What am I missing?

Comment (wxiaoguang) [reply modules/indexer/issues/indexer.go]: Have you refreshed your search indexer? Maybe you can dump the document in the search engine to see whether they are correct.

**Comment by aceArt-GmbH** [comment]
It is working now :D
(although I still don't know how to refreshed the search indexer)

I didn't test the elastic search part.

Please test and review as I am not a Go dev

**Comment by aceArt-GmbH** [comment]
is it normal for testing-amd64 test-mysql to fail?
What is the problem?

**Comment by wxiaoguang** [comment]
> It is working now :D (although I still don't know how to refreshed the search indexer)
> 
> I didn't test the elastic search part.
> 
> Please test and review as I am not a Go dev

You can make some edits to an issue, then the search indexer will update the content.

ps: I think it's better to have some unit test code to guarantee the code works as expected. If you have any question, Gitea maintainers will help.

> is it normal for testing-amd64 test-mysql to fail?

Some failures are caused by some bugs in Gitea test code, it doesn't matter.

**Comment by aceArt-GmbH** [comment]
>I think it's better to have some unit test code to guarantee the code works as expected.

I added a case to bleve_test.go, is that enough?

**Comment by wxiaoguang** [comment]
That's fine. I just realized that there was no unit test for elasticsearch ...

File: modules/indexer/issues/bleve.go [hunk 2]
```diff
@@ -236,6 +239,7 @@ func (b *BleveIndexer) Search(keyword string, repoIDs []int64, limit, start int)
 	for _, repoID := range repoIDs {
 		repoQueriesP = append(repoQueriesP, numericEqualityQuery(repoID, "RepoID"))
 	}
+	index, _ := strconv.ParseInt(keyword, 10, 64)
```
Comment (lunny) [inline modules/indexer/issues/bleve.go:250]: Maybe keyword is a string?

**Comment by lunny** [comment]
To keep the compatible, we need to upgrade indexer version so that all index will be rebuilt.

File: modules/indexer/issues/bleve.go (hunk 2 above)
Comment (aceArt-GmbH) [reply modules/indexer/issues/bleve.go:250]: does it matter?
I didn't notice a difference in behavior.
How would I dynamically add the query if keyword is a string and the parser errors?

**Comment by aceArt-GmbH** [comment]
> To keep the compatible, we need to upgrade indexer version so that all index will be rebuilt.

How do I do that? Increase `issueIndexerLatestVersion`?

**Comment by lunny** [comment]
> > To keep the compatible, we need to upgrade indexer version so that all index will be rebuilt.
> 
> How do I do that? Increase `issueIndexerLatestVersion`?

Yes

**Comment by aceArt-GmbH** [comment]
I increased the `issueIndexerLatestVersion`.
Anything else left here? :)

**Review by wxiaoguang — APPROVED** [review]

**Comment by zeripath** [comment]
Why did you close this?

**Review by zeripath — APPROVED** [review]

**Comment by aceArt-GmbH** [comment]
> Why did you close this?

Because this is taking too long

**Comment by 6543** [comment]
> Because this is taking too long

consider that maintainers maintain this project in there free time, so yes it can take long - sometimes ~2years++ for big features.

**Comment by 6543** [comment]
so in your case it just fall out of the "manual merge queue" i guess

**Comment by 6543** [comment]
@zeripath @wxiaoguang @aceArt-GmbH  ->  #19646

**Comment by 6543** [comment]
> so in your case it just fall out of the "manual merge queue" i guess

I think this is because github does not allow maintainers to "update pulls" because you created a pull from an repo owned by an organisation ...

**Pull request closed without merging**